File: backend/src/routes/shares.ts
import { Router, Request, Response, NextFunction } from 'express';
import multer from 'multer';
import { z } from 'zod';
import * as ShareService from '../services/ShareService';
import { isValidUID, normalizeUID, formatUID } from '../services/UIDService';
import { verifyPassword, createSessionToken, validateSessionToken } from '../services/PasswordService';
import { uploadLimiter, retrievalLimiter, passwordLimiter } from '../middleware/rateLimiter';
import { mimeValidator } from '../middleware/mimeValidator';
import { validateBody } from '../middleware/validateBody';
import { createError } from '../middleware/errorHandler';
import { env } from '../config/env';

const router = Router();

// ─── Multer config ──────────────────────────────────────
const upload = multer({
  storage: multer.memoryStorage(),
  limits: {
    fileSize: env.MAX_UPLOAD_BYTES,
    files: 20, // max 20 files per upload
  },
});

// ─── Validation schemas ─────────────────────────────────
const textShareSchema = z.object({
  title: z.string().max(200).optional(),
  content: z.string().min(1, 'Content is required').max(500000, 'Content too large (max 500KB)'),
  language: z.string().max(50).optional(),
  expiresIn: z.enum(['1h', '6h', '24h', '7d', '30d']).optional(),
  password: z.string().min(4).max(128).optional(),
});

const verifySchema = z.object({
  password: z.string().min(1, 'Password is required'),
});

// ─── POST /api/shares/file — Create file share ─────────
router.post(
  '/file',
  uploadLimiter,
  upload.array('files', 20),
  mimeValidator,
  async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
      const files = req.files as Express.Multer.File[];
      if (!files || files.length === 0) {
        throw createError('No files uploaded', 400);
      }

      // Calculate total size
      const totalSize = files.reduce((sum, f) => sum + f.size, 0);
      if (totalSize > env.MAX_UPLOAD_BYTES) {
        throw createError(`Total upload size exceeds ${env.MAX_UPLOAD_BYTES / (1024 * 1024)}MB limit`, 400);
      }

      const input: ShareService.CreateFileShareInput = {
        files: files.map((f) => ({
          filename: f.originalname,
          buffer: f.buffer,
          mimeType: f.mimetype,
          size: f.size,
        })),
        expiresIn: req.body.expiresIn,
        password: req.body.password,
      };

      const uid = await ShareService.createFileShare(input);

      res.status(201).json({
        success: true,
        data: {
          uid,
          formattedUID: formatUID(uid),
        },
      });
    } catch (error) {
      next(error);
    }
  }
);

// ─── POST /api/shares/text — Create text share ─────────
router.post(
  '/text',
  uploadLimiter,
  validateBody(textShareSchema),
  async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
      const uid = await ShareService.createTextShare(req.body);

      res.status(201).json({
        success: true,
        data: {
          uid,
          formattedUID: formatUID(uid),
        },
      });
    } catch (error) {
      next(error);
    }
  }
);

// ─── GET /api/shares/:uid — Get share metadata ─────────
router.get(
  '/:uid',
  retrievalLimiter,
  async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
      const uid = normalizeUID(req.params.uid);
      if (!isValidUID(uid)) {
        throw createError('Invalid UID format — must be 12 digits', 400);
      }

      const share = await ShareService.getShareByUID(uid);
      if (!share) {
        throw createError('Share not found or has expired', 404);
      }

      res.json({
        success: true,
        data: share,
      });
    } catch (error) {
      next(error);
    }
  }
);

// ─── POST /api/shares/:uid/verify — Password verify ────
router.post(
  '/:uid/verify',
  passwordLimiter,
  validateBody(verifySchema),
  async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
      const uid = normalizeUID(req.params.uid);
      if (!isValidUID(uid)) {
        throw createError('Invalid UID format', 400);
      }

      const hash = await ShareService.getSharePasswordHash(uid);
      if (!hash) {
        throw createError('Share not found or is not password-protected', 404);
      }

      const isValid = await verifyPassword(req.body.password, hash);
      if (!isValid) {
        throw createError('Incorrect password', 401);
      }

      // Create a short-lived session token
      const sessionToken = createSessionToken(uid);

      res.json({
        success: true,
        data: { sessionToken },
      });
    } catch (error) {
      next(error);
    }
  }
);

// ─── GET /api/shares/:uid/content — Get text content ───
router.get(
  '/:uid/content',
  retrievalLimiter,
  async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
      const uid = normalizeUID(req.params.uid);
      if (!isValidUID(uid)) {
        throw createError('Invalid UID format', 400);
      }

      // Check if share exists and if auth is needed
      const share = await ShareService.getShareByUID(uid);
      if (!share) {
        throw createError('Share not found or has expired', 404);
      }

      // If private, validate session token
      if (share.isPrivate) {
        const token = req.headers['x-session-token'] as string;
        if (!token || !validateSessionToken(token, uid)) {
          throw createError('Authentication required — verify password first', 401);
        }
      }

      const content = await ShareService.getTextContent(uid);
      if (!content) {
        throw createError('Text content not found', 404);
      }

      res.json({
        success: true,
        data: content,
      });
    } catch (error) {
      next(error);
    }
  }
);

export default router;

File: backend/app/routers/shares.py
"""
shares.py — Share creation, retrieval, password verification, and text content routes.
Replaces: backend/src/routes/shares.ts
"""

from fastapi import APIRouter, Depends, Request, UploadFile, Form, Header
from fastapi.responses import JSONResponse
from sqlalchemy.ext.asyncio import AsyncSession
from pydantic import ValidationError

from app.database import get_db
from app.schemas import TextShareCreate, PasswordVerify
from app.services import share_service, password_service
from app.services.uid_service import normalize_uid, is_valid_uid, format_uid
from app.middleware.rate_limiter import limiter, RETRIEVAL_LIMIT, PASSWORD_LIMIT, UPLOAD_LIMIT
from app.middleware.mime_validator import validate_file
from app.config import settings

router = APIRouter(prefix="/api/shares")


# ─── POST /api/shares/file — Create file share ─────────


@router.post("/file")
@limiter.limit(UPLOAD_LIMIT)
async def create_file_share(
    request: Request,
    files: list[UploadFile] = [],
    expiresIn: str | None = Form(default=None),
    password: str | None = Form(default=None),
    db: AsyncSession = Depends(get_db),
):
    if not files:
        return JSONResponse(
            status_code=400,
            content={"success": False, "error": "No files uploaded"},
        )

    if len(files) > 20:
        return JSONResponse(
            status_code=400,
            content={"success": False, "error": "Maximum 20 files per upload"},
        )

    # Read all file contents and validate
    file_data = []
    total_size = 0
    for f in files:
        content = await f.read()
        total_size += len(content)

        if total_size > settings.MAX_UPLOAD_BYTES:
            return JSONResponse(
                status_code=400,
                content={
                    "success": False,
                    "error": f"Total upload size exceeds {settings.MAX_UPLOAD_BYTES // (1024 * 1024)}MB limit",
                },
            )

        # MIME validation via magic bytes
        mime_error = validate_file(f.filename or "unknown", content)
        if mime_error:
            return JSONResponse(
                status_code=400,
                content={"success": False, "error": mime_error},
            )

        file_data.append(
            {
                "filename": f.filename or "unknown",
                "buffer": content,
                "mime_type": f.content_type or "application/octet-stream",
                "size": len(content),
            }
        )

    uid = await share_service.create_file_share(
        db, file_data, expires_in=expiresIn, password=password
    )

    return JSONResponse(
        status_code=201,
        content={
            "success": True,
            "data": {"uid": uid, "formattedUID": format_uid(uid)},
        },
    )


# ─── POST /api/shares/text — Create text share ─────────


@router.post("/text")
@limiter.limit(UPLOAD_LIMIT)
async def create_text_share(
    request: Request,
    body: TextShareCreate,
    db: AsyncSession = Depends(get_db),
):
    uid = await share_service.create_text_share(
        db,
        content=body.content,
        title=body.title,
        language=body.language,
        expires_in=body.expiresIn,
        password=body.password,
    )

    return JSONResponse(
        status_code=201,
        content={
            "success": True,
            "data": {"uid": uid, "formattedUID": format_uid(uid)},
        },
    )


# ─── GET /api/shares/{uid} — Get share metadata ────────


@router.get("/{uid}")
@limiter.limit(RETRIEVAL_LIMIT)
async def get_share(
    request: Request,
    uid: str,
    db: AsyncSession = Depends(get_db),
):
    clean_uid = normalize_uid(uid)
    if not is_valid_uid(clean_uid):
        return JSONResponse(
            status_code=400,
            content={"success": False, "error": "Invalid UID format — must be 12 digits"},
        )

    share = await share_service.get_share_by_uid(db, clean_uid)
    if not share:
        return JSONResponse(
            status_code=404,
            content={"success": False, "error": "Share not found or has expired"},
        )

    return {"success": True, "data": share}


# ─── POST /api/shares/{uid}/verify — Password verify ───


@router.post("/{uid}/verify")
@limiter.limit(PASSWORD_LIMIT)
async def verify_share_password(
    request: Request,
    uid: str,
    body: PasswordVerify,
    db: AsyncSession = Depends(get_db),
):
    clean_uid = normalize_uid(uid)
    if not is_valid_uid(clean_uid):
        return JSONResponse(
            status_code=400,
            content={"success": False, "error": "Invalid UID format"},
        )

    pw_hash = await share_service.get_share_password_hash(db, clean_uid)
    if not pw_hash:
        return JSONResponse(
            status_code=404,
            content={"success": False, "error": "Share not found or is not password-protected"},
        )

    is_valid = await password_service.verify_password(body.password, pw_hash)
    if not is_valid:
        return JSONResponse(
            status_code=401,
            content={"success": False, "error": "Incorrect password"},
        )

    session_token = password_service.create_session_token(clean_uid)

    return {"success": True, "data": {"sessionToken": session_token}}


# ─── GET /api/shares/{uid}/content — Get text content ──


@router.get("/{uid}/content")
@limiter.limit(RETRIEVAL_LIMIT)
async def get_text_content(
    request: Request,
    uid: str,
    db: AsyncSession = Depends(get_db),
    x_session_token: str | None = Header(default=None, alias="X-Session-Token"),
):
    clean_uid = normalize_uid(uid)
    if not is_valid_uid(clean_uid):
        return JSONResponse(
            status_code=400,
            content={"success": False, "error": "Invalid UID format"},
        )

    share = await share_service.get_share_by_uid(db, clean_uid)
    if not share:
        return JSONResponse(
            status_code=404,
            content={"success": False, "error": "Share not found or has expired"},
        )

    # If private, validate session token
    if share["isPrivate"]:
        if not x_session_token or not password_service.validate_session_token(
            x_session_token, clean_uid
        ):
            return JSONResponse(
                status_code=401,
                content={
                    "success": False,
                    "error": "Authentication required — verify password first",
                },
            )

    content = await share_service.get_text_content(db, clean_uid)
    if not content:
        return JSONResponse(
            status_code=404,
            content={"success": False, "error": "Text content not found"},
        )

    return {"success": True, "data": content}
